OSDOCS-2911: Adding documentation about conditional updates

sagidlow opened the pull request:
Applies to 4.10+
JIRA: https://issues.redhat.com/browse/OSDOCS-2911
SME and QE ack required.
Preview Links: 

- [Conditional updates](https://deploy-preview-40412--osdocs.netlify.app/openshift-enterprise/latest/updating/understanding-upgrade-channels-release.html#conditional-updates-overview_understanding-upgrade-channels-releases) in the Understanding upgrade channels and releases section. 
- [Updating to a conditional upgrade path](https://deploy-preview-40412--osdocs.netlify.app/openshift-enterprise/latest/updating/updating-cluster-cli.html#update-conditional-upgrade-pathupdating-cluster-cli) in the Updating a cluster within a minor version using the CLI section.

sagidlow commented:
@sdodson, @wking and @LalatenduMohanty please review for production documentation for OSDOCS-2911.

wking commented on modules/understanding-upgrade-channels.adoc:
```diff
@@ -99,6 +99,32 @@ endif::openshift-origin[]
 
```
do we want to fiddle with "after the update recommendation has been removed" a few lines up^^?  Maybe "after the update recommendation has been removed or been made conditional" or some such?

sagidlow replied:
My only hesitation with this was that we were going to drop new content in the middle of a section. So, I figured at least its own section would be better and would be easier for linking and standing out.

wking replied:
I like having the new section.  I just think that "after the update recommendation has been removed" is a bit stale in this new context, and that we probably want to expand it so that folks know when they're supported.  Cases:

* Update is currently recommended for your cluster: definitely supported.
* Update was previously recommended, but is now completely gone from the recommended-update chart: still supported, but unlikely to happen anymore with 4.10 -> * updates, because we'll use a conditional edge.
* Update was previously recommended, but is now conditionally recommended, and for the cluster you're looking at, a risk applies, and it's not recommended: still supported.  This is the bit I think we want to call out new in this PR.
* Update was never recommended in any supported channel: not supported.

sagidlow replied:
> ifdef::openshift-origin[]
The presence of an update recommendation in the `stable-4` channel at any point is a declaration that the update is supported. While releases will never be removed from the channel, update recommendations that exhibit serious issues will be removed from the channel. Updates initiated after the update recommendation has been removed are still supported.
endif::openshift-origin[]

Update to:

ifdef::openshift-origin[]
The presence of an update recommendation in the `stable-4` channel at any point is a declaration that the update is supported. While releases will never be removed from the channel, update recommendations that exhibit serious issues will be removed from the channel. Update that was previously recommended, but is now conditionally recommended, and for the cluster, you wish to update, a risk still applies, and it is not recommended for an update is supported. 
endif::openshift-origin[]

wking replied:
> ...update recommendations that exhibit serious issues will be removed from the channel...

Ah, this is going to be stale too.  Now it will be "may be removed from the channel, or may become conditional updates", or some such.

> Update that was previously...

"An update that..."?  Also, the rest of this sentence has a lot of conditional comments.  The main point is that if the update was ever recommended in a supported channel for any cluster will always be supported for every cluster.  What if they become conditional?  Still supported.  What if they are removed?  Still supported.  Maybe something like:

    Once an update recommendation is supported, it remains supported for the life of {product-version}, even if the update recommendation is later dropped or made conditional.

Also, looks like there is similar wording for OCP and OKD, and we should probably update both, although I'm a bit fuzzy on what "supported" means for OKD.

```console
$ git grep -n 'The presence of an update recommendation' | cut -b -180
modules/understanding-upgrade-channels.adoc:91:Update stability depends on your channel. The presence of an update recommendation in the `candidate-{product-version}` channel does 
modules/understanding-upgrade-channels.adoc:97:The presence of an update recommendation in the `stable-4` channel at any point is a declaration that the update is supported. While 
```

sagidlow replied:
Done

wking commented on updating/updating-cluster-cli.adoc:
```diff
@@ -43,4 +43,8 @@ For information on which machine configuration changes require a reboot, see the
 
+.Additional resources
+
+* xref:../updating/understanding-upgrade-channels-release.adoc#update-conditional-updates[Conditional updates]
```
And perhaps these `Additional resources` links should just to `../updating/understanding-upgrade-channels-release.adoc` as a whole (and maybe they already do?) for "you should understand how channels and update paths work before actually attempting an update"?

sagidlow replied:
Actually, I agree that we should link to upgrade channels as a whole instead of just the conditional. I'll update them here shortly.

sagidlow replied:
@wking updating/updating-cluster.adoc has been removed and replaced with updating/updating-cluster-within-minor.adoc

LalatenduMohanty reviewed (approved):
/lgtm

shellyyang1989 commented:
The overview looks good to me. But in addition, we need to add details to instruct how users can update to conditional updates in `Updating a cluster within a minor version using the CLI`.

wking commented:
> But in addition, we need to add details to instruct how users can update to conditional updates in Updating a cluster within a minor version using the CLI.

Hmm, good point.  We are currently adding both context and mechanics in the context-focused doc.  But neither the context nor the `oc ...` oneliners are particularly complicated, so if we tried to keep the context in `understanding-upgrade-channels.adoc` and the mechanics in `updating/updating-cluster-cli.adoc`, etc., I think we'd end up feeling like "ok, but how?" in the context section, and "ok, but why?" in the mechanics section.  Not clear to me how to efficiently resolve this issue...

sagidlow commented:
@shellyyang1989 @wking - What about also adding this content in the `updating/updating-cluster-cli.adoc`?

sagidlow commented:
@shellyyang1989 and @wking okay I added a new module for conditional updates instead of a new section in the channels chapter. I had to move the placement of the Conditional updates in the Channels section because I can't include a module in a module. I also added the same module in the `updating/updating-cluster-cli.adoc` this way we have 1 place to make updates and not several. It'll make maintenance easier. If this doesn't work let me know.

shellyyang1989 commented:
> What about also adding this content in the updating/updating-cluster-cli.adoc?
From my point of view, the `Understanding upgrade channels and releases` introduces what upgrade channels are, as well as upgrade paths. It's kind of high level. So I think we can put the introduction of conditional updates here. While  `Updating a cluster within a minor version using the CLI` instructs users how to upgrade a cluster using cli. It's talking about the detailed procedures. So we can merge the procedures of updating to conditional edges into the section. Complete copy about conditional edge in the two sections looks not that good to users.

sagidlow commented:
@shellyyang1989 you're suggesting leaving this as a part of the `Understanding upgrade channels and releases`

> The OpenShift Update Service may declare conditionally recommended updates associated with known risks.
>  The Cluster Version Operator (CVO) continually evaluates known risks associated with updates from the current {product-title} release to later releases.
>  When no risks apply, the update is recommended and are available.

And leaving the procedural instructions for `Updating a cluster within a minor version using the CLI`?

wking commented on modules/update-conditional-updates.adoc:
```diff
@@ -0,0 +1,39 @@
+// * updating/understanding-upgrade-channels-releases.adoc
+
+[id="update-conditional-upgrade-path{context}"]
+= Updating to a conditional upgrade path
```
"to an ... upgrade path" -> "along an ... upgrade path"?  Also "update along a conditional upgrade path" and such below?

wking commented on modules/update-conditional-updates.adoc:
```diff
@@ -0,0 +1,39 @@
+[id="update-conditional-upgrade-path{context}"]
+= Updating to a conditional upgrade path
+
+You can update a conditional upgrade path using the OpenShift CLI (`oc`) 4.10 or later.
```
Possibly worth pointing out again here that when a conditional update happens to be recommended for your cluster, that you can update using the same web console and `oc` commands that folks have been using since 4.1.  It's only when a conditional update is not recommended for your cluster that you want the 4.10+ `oc`.

shellyyang1989 replied:
Yeah, agree. And if we pointing it out by wording, I think we can remove line#13 through line#21.

shellyyang1989 commented on modules/understanding-upgrade-channels.adoc:
```diff
@@ -88,13 +88,13 @@ endif::openshift-origin[]
 
 ifndef::openshift-origin[]
-Update stability depends on your channel. The presence of an update recommendation in the `candidate-{product-version}` channel does not imply that the update is supported. It means that no serious issues have been found with the update yet, but there might not be significant traffic through the update to suggest stability. The presence of an update recommendation in the `fast-{product-version}`  ...
+Update stability depends on your channel. The presence of an update recommendation in the `candidate-{product-version}` channel does not imply that the update is supported. It means that no serious issues have been found with the update yet, but there might not be significant traffic through the update to suggest stability. The presence of an update recommendation in the `fast-{product-version}`  ...
```
> update recommendations that exhibit serious issues will be removed from all channels

Changing to `update recommendations that exhibit serious issues will be removed or made conditional from all channels` makes sense?

wking replied:
yeah, and maybe drop the `Updates initiated after the update recommendation has been removed are still supported.` sentence, since that aspect is now more generally covered in the new `Once an update recommendation is supported` sentence.

shellyyang1989 commented on modules/understanding-upgrade-channels.adoc:
```diff
@@ -88,13 +88,13 @@ endif::openshift-origin[]
 
 ifdef::openshift-origin[]
-The presence of an update recommendation in the `stable-4` channel at any point is a declaration that the update is supported. While releases will never be removed from the channel, update recommendations that exhibit serious issues will be removed from the channel. Updates initiated after the update recommendation has been removed are still supported.
+The presence of an update recommendation in the `stable-4` channel at any point is a declaration that the update is supported. While releases will never be removed from the channel, update recommendations that exhibit serious issues will be removed from the channel. Updates initiated after the update recommendation has been removed are still supported. Once an update recommendation is supported,  ...
```
Same here.
Changing to `update recommendations that exhibit serious issues will be removed or made conditional from all channels` makes sense?

wking commented on modules/update-conditional-updates.adoc:
```diff
@@ -0,0 +1,39 @@
++
+[source,terminal]
+----
+$ oc adm upgrade --to <version> <.>
```
We cover this in `modules/update-upgrading-cli.adoc`.  Possibly worth scoping this module down to just the "...but what about when a conditional update is not recommended?" case?

wking replied:
Hmm, I'm a bit confused by the web vs. `oc` breakdown and the patch vs. minor breakdown:

```console
$ git --no-pager grep modules/update-upgrading origin/main --
origin/main:updating/updating-cluster-cli.adoc:include::modules/update-upgrading-cli.adoc[leveloffset=+1]
origin/main:updating/updating-cluster-rhel-compute.adoc:include::modules/update-upgrading-web.adoc[leveloffset=+1]
origin/main:updating/updating-cluster-within-minor.adoc:include::modules/update-upgrading-web.adoc[leveloffset=+1]
origin/main:updating/updating-cluster.adoc:include::modules/update-upgrading-web.adoc[leveloffset=+1]
```

I would expect all combinations (web patch, `oc` patch, web minor, `oc` minor), because the steps that you take for a patch update all happen for minor updates too.  Not sure how much of that we want to involve ourselves in in this PR.

shellyyang1989 commented on modules/update-conditional-updates.adoc:
```diff
@@ -0,0 +1,39 @@
+// * updating/understanding-upgrade-channels-releases.adoc
+
+[id="update-conditional-upgrade-path{context}"]
+= Updating to a conditional upgrade path
```
Changing to `Updating to a conditional update` is better?

sagidlow replied:
`Updating to a conditional update` is a bit repetitive. Changing to `Updating along a conditional upgrade path`.

shellyyang1989 commented on modules/update-conditional-updates.adoc:
```diff
@@ -0,0 +1,39 @@
+The recommended updates are also available on the web console from the Administrator perspective.
+When updates are not recommended because a risk might apply you can view the view the description of the update.
++
+To view the description of the update, run the following command:
```
Possibly worth simplifying to:

1. When updates are not recommended because a risk might apply, view the description of the update:
    $ oc adm upgrade --include-not-recommended

sagidlow replied:
There is a convention to use ", run the following command." in docs. So, I'm making an update as recommended with the convention.

shellyyang1989 commented on modules/update-conditional-updates.adoc:
```diff
@@ -0,0 +1,39 @@
+. If the cluster administrator evaluates the potential known risks and decides it is acceptable for the current cluster, then the administrator can waive the safety guards.
++
+To waive the safety guards and continue the update, run the following command:
++
```
Possibly worth simplifying to:

2. If the cluster administrator evaluates the potential known risks and decides it is acceptable for the current cluster, then the administrator can waive the safety guards and proceed the update:
   $ oc adm upgrade --allow-not-recommended --to <version>

shellyyang1989 commented on modules/update-conditional-updates.adoc:
```diff
@@ -0,0 +1,39 @@
+----
+$ oc adm upgrade --allow-not-recommended --to <version> <.>
+----
+<.> `<version>` is the update version you want to update to.
```
<version> is the supported but not recommended update version that you obtained from the output of the previous command.

sagidlow commented:
@shellyyang1989 @wking PTAL.

wking reviewed (approved):
/lgtm

shellyyang1989 commented:
lgtm.
One concern, since we name the section to Updating along a conditional update path, to make it consistent, do we need to change Updating a cluster by using the CLI to Updating along a recommended update path?

sagidlow commented:
@shellyyang1989 Would prefer to keep it as is since there is a web console version. It'll keep it consistent :) i.e., updating using the CLI and then updating using the web console

stevsmit commented on modules/understanding-upgrade-channels.adoc:
```diff
@@ -88,13 +88,13 @@ endif::openshift-origin[]
 
 ifndef::openshift-origin[]
-Update stability depends on your channel. The presence of an update recommendation in the `candidate-{product-version}` channel does not imply that the update is supported. It means that no serious issues have been found with the update yet, but there might not be significant traffic through the update to suggest stability. The presence of an update recommendation in the `fast-{product-version}`  ...
+Update stability depends on your channel. The presence of an update recommendation in the `candidate-{product-version}` channel does not imply that the update is supported. It means that no serious issues have been found with the update yet, but there might not be significant traffic through the update to suggest stability. The presence of an update recommendation in the `fast-{product-version}`  ...
```
s/ once/ When based on IBMSG? https://docs.engineering.redhat.com/display/CCSENABLE/Avoid+these+words

stevsmit commented on modules/understanding-upgrade-channels.adoc:
```diff
@@ -88,13 +88,13 @@ endif::openshift-origin[]
 
 ifdef::openshift-origin[]
-The presence of an update recommendation in the `stable-4` channel at any point is a declaration that the update is supported. While releases will never be removed from the channel, update recommendations that exhibit serious issues will be removed from the channel. Updates initiated after the update recommendation has been removed are still supported.
+The presence of an update recommendation in the `stable-4` channel at any point is a declaration that the update is supported. While releases will never be removed from the channel, update recommendations that exhibit serious issues will be removed or made conditional from all channels. Once an update recommendation is supported, it remains supported for the life of {product-version}, even if the ...
```
Same thing as above^

stevsmit commented on modules/understanding-upgrade-channels.adoc:
```diff
@@ -136,3 +136,11 @@ Changing your channel might impact the supportability of your cluster. The follo
+[id="conditional-updates-overview_{context}"]
+== Conditional updates
+
+The OpenShift Update Service may declare conditionally recommended updates associated with known risks.
```
s/ may/ might https://docs.engineering.redhat.com/display/CCSENABLE/Avoid+these+words

stevsmit commented on updating/understanding-upgrade-channels-release.adoc:
```diff
@@ -33,3 +33,6 @@ ifdef::openshift-origin[]
 include::modules/understanding-upgrade-channels.adoc[leveloffset=+1]
+
+.Additional resources
+* For more information, see xref:../updating/updating-cluster-cli.adoc#update-conditional-upgrade-pathupdating-cluster-cli[Updating to a conditional upgrade path]
```
The title of the linked page is "Updating along a conditional upgrade path", but the ID is "Updating to a conditional upgrade path." Not sure if it needs changed, just FYI!

stevsmit commented on updating/updating-cluster-cli.adoc:
```diff
@@ -42,5 +42,10 @@ include::modules/updating-sno.adoc[leveloffset=+1]
+
+.Additional resources
+
+* xref:../updating/understanding-upgrade-channels-release.adoc#understanding-upgrade-channels_understanding-upgrade-channels-releases[Understanding upgrade channels and releases]
```
Linked page is "Upgrade channels and release paths", and link ID here is "Understanding upgrade channels and releases." Not sure if matters (inclined to say they should match, though)

stevsmit commented:
/lgtm

shellyyang1989 commented:
> Would prefer to keep it as is since there is a web console version. It'll keep it consistent :) i.e., updating using the CLI and then updating using the web console

I do agree to keep the section title as it is: `Updating a cluster within a minor version using the CLI` which is consistent with `Updating a cluster within a minor version using the web console`.
However, in the section `Updating a cluster within a minor version using the CLI`, I feel the sub-sections should be consistent as well. If we have `Updating along a conditional update path` as a sub-section name, changing the existing to `Updating along a recommended update path` looks more consistent.

shellyyang1989 commented:
One more, the example output of  `oc adm upgrade` in section `Updating a cluster by using the CLI` needs update by changing `Updates` to `Recommended updates`. I do have Upstream and Channel displayed in the output. Would you like to get them added?

ex:
```
# oc adm upgrade 
Cluster version is 4.10.0-0.nightly-2022-01-11-065245

Upstream: https://openshift-release.apps.ci.l2s4.p1.openshiftapps.com/graph
Channel: stable-4.10

Recommended updates:

  VERSION                            IMAGE
  4.10.0-0.nightly-2022-01-13-061145 registry.ci.openshift.org/ocp/release@sha256:c03f58e1f3000ddeabcfeb61f343fba640e9354e2e5347c044b41edb4faf7b96
```

sagidlow commented:
/cherrypick enterprise-4.10
